examples/sha256_nivc.rs

```diff
@@ -0,0 +1,222 @@
+        let mut hasher = Sha256::new();
+
+        let mut input = vec![0u8; 64];
+
+        for (i, input_ptr) in args.iter().enumerate() {
+            let input_zptr = s.hash_expr(input_ptr).unwrap();
+            let tag_zptr: F = input_zptr.tag().to_field();
+            let hash_zptr = input_zptr.value();
+            input[(64 * i)..(64 * i + 32)].copy_from_slice(&tag_zptr.to_bytes());
+            input[(64 * i + 32)..(64 * (i + 1))].copy_from_slice(&hash_zptr.to_bytes());
+        }
+
+        input.reverse();
+
+        hasher.update(input);
+        let mut bytes = hasher.finalize();
+        bytes.reverse();
+        let l = bytes.len();
+        // Discard the two most significant bits.
+        bytes[l - 1] &= 0b00111111;
+
+        let scalar = F::from_bytes(&bytes).unwrap();
+        let result = Num::from_scalar(scalar);
+
+        s.intern_num(result)
+    }
+
+    fn has_circuit(&self) -> bool {
+        true
+    }
+}
+
+impl<F: LurkField> Sha256Coprocessor<F> {
+    pub(crate) fn new() -> Self {
+        Self {
+            _p: Default::default(),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Coproc, Serialize, Deserialize)]
+enum Sha256Coproc<F: LurkField> {
+    SC(Sha256Coprocessor<F>),
+}
+
+/// Run the example in this file with
+/// `cargo run --release --example sha256_nivc`
+/// where `n` is the needed arity
+fn main() {
+    let subscriber = Registry::default()
+        .with(fmt::layer().pretty())
+        .with(EnvFilter::from_default_env())
+        .with(TeXRayLayer::new());
+    tracing::subscriber::set_global_default(subscriber).unwrap();
+
+    let store = &mut Store::<Fr>::new();
+    let cproc_sym = user_sym("sha256");
+
+    let call = sha256_encode(store);
+
+    let lang = Lang::<Fr, Sha256Coproc<Fr>>::new_with_bindings(
+        store,
+        vec![(cproc_sym, Sha256Coprocessor::new().into())],
+    );
+    let lang_rc = Arc::new(lang.clone());
+
+    let supernova_prover = SuperNovaProver::<Fr, Sha256Coproc<Fr>>::new(REDUCTION_COUNT, lang);
+
+    // println!("Setting up public parameters (rc = {REDUCTION_COUNT})...");
+
+    let pp_start = Instant::now();
+
+    // // see the documentation on `with_public_params`
+    // with_public_params(REDUCTION_COUNT, lang_rc.clone(), |pp| {
+    let pp_end = pp_start.elapsed();
+    // println!("Public parameters took {:?}", pp_end);
+
+    println!("Beginning proof+public-parameters step...");
+    let proof_start = Instant::now();
+    let (_proof, _z0, _zi, _num_steps) = supernova_prover
+        .evaluate_and_prove(None, call, empty_sym_env(store), store, 10000, lang_rc)
+        .unwrap();
+    let proof_end = proof_start.elapsed();
+
+    println!("Proofs took {:?}", proof_end);
+
+    // TODO: plumb verification.
+    // println!("Verifying proof...");
+
+    // let verify_start = Instant::now();
+    // let res = proof.verify(claim, &z0, &zi).unwrap();
+    // let verify_end = verify_start.elapsed();
+
+    // println!("Verify took {:?}", verify_end);
```

Comment:
Yes, it would. It's not entirely obvious what the right interface for this will be. Whereas a Nova `RecursiveSnark` contains the `z_i` needed to verify itself, a SuperNova `RecursiveSnark` does not. It needs to also be provided with a `RunningClaim`. In order to provide an interface as much like Nova's as possible, it may be necessary to refactor SuperNova in arecibo. @mpenciak is looking into this. I'll create an issue to track this understanding, and if we end up deciding to hack around it in Lurk, we can. For now, I assume the right thing is to reshape the upstream interfaces.